nifi-nar-bundles/nifi-hive-bundle/nifi-hive-processors/src/main/java/org/apache/nifi/processors/hive/PutHiveStreaming.java

```diff
@@ -0,0 +1,657 @@
+                partitionColumnList = Collections.emptyList();
+            } else {
+                String[] partitionCols = partitionColumns.split(",");
+                partitionColumnList = new ArrayList<>(partitionCols.length);
+                for (String col : partitionCols) {
+                    partitionColumnList.add(col.trim());
+                }
+            }
+
+            // Store the original class loader, then explicitly set it to this class's classloader (for use by the Hive Metastore)
+            ClassLoader originalClassloader = Thread.currentThread().getContextClassLoader();
+            Thread.currentThread().setContextClassLoader(this.getClass().getClassLoader());
+
+            int recordCount = 0;
+            final List<HiveStreamingRecord> records = new LinkedList<>();
+
+            session.read(flowFile, in -> {
+
+                try (final DataFileStream<GenericRecord> reader = new DataFileStream<>(in, new GenericDatumReader<GenericRecord>())) {
+
+                    GenericRecord currRecord;
+                    while (reader.hasNext()) {
+                        currRecord = reader.next();
+                        List<String> partitionValues = new ArrayList<>();
+
+                        for (String partition : partitionColumnList) {
+                            Object partitionValue = currRecord.get(partition);
+                            if (partitionValue == null) {
+                                throw new IOException("Partition column '" + partition + "' not found in Avro record");
+                            }
+                            partitionValues.add(partitionValue.toString());
+                        }
+
+                        List<Schema.Field> fields = currRecord.getSchema().getFields();
+                        if (fields != null) {
+                            JSONObject obj = new JSONObject();
+                            for (Schema.Field field : fields) {
+                                String fieldName = field.name();
+                                // Skip fields that are partition columns, we extracted those values above to create an EndPoint
+                                if (!partitionColumnList.contains(fieldName)) {
+                                    Object value = currRecord.get(fieldName);
+                                    try {
+                                        obj.put(fieldName, value);
+                                    } catch (JSONException je) {
+                                        throw new IOException(je);
+                                    }
+                                }
+                            }
+                            records.add(new HiveStreamingRecord(partitionValues, obj));
+                        }
+                    }
+                }
+            });
+
+            // Write all records to Hive Streaming
+            for (HiveStreamingRecord record : records) {
```

Comment:
I don't know much about how the HiveWriter works, but would there be any benefit to writing the records as you go in the above loop, rather than reading them all into memory and then writing them? 

If we leave as is maybe we can just put a note in the processor description that mentions that the full contents of the incoming Avro will be read into memory, or if this processor/Hive wouldn't perform well with large Avro files then maybe we should mention that instead.